### PR TITLE
fixed mask borders

### LINKED ISSUE #10330: Instance segmentation: quick but not good enough
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Thanks for your contribution! I have tested the instance segmentation demo, but the results were not good, people's heads always not fully segment, like yolact. I also tried padding, and the same results were presented.
Here are some comparison：
yolov5x-seg result
Capture
yolov7 result
Capturev7
IDOL result
Capture_IDOL

### Additional

_No response_

### PR DESCRIPTION
**Pull request by paradigmn**
The masks appear to be clipped incorrectly under the bounding boxes. The crop_mask() function determines the mask dimensions by comparing the predicted box coordinates to incrementing integer vectors in order to determine the pixel borders. However, this approach will always choose the next greater pixel (roof function), if the the predicted value is minimal higher than the corresponding integer value. The mask upscaling magnifies this issue by a factor of four. Rounding before the comparison seems to alleviate the issue.

Additionally, I fixed the documentation of the vector dimensions.

Below you see the before and after compassion for the s-model (see car on the left):
tampere
tampere2

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improvement to segmentation mask cropping logic.

### 📊 Key Changes
- Adjusted the expected shape of mask tensors from `[h, w, n]` to `[n, h, w]`.
- Modified indexing calculations to use center points in bounding box comparisons for tensor shapes and device allocation.

### 🎯 Purpose & Impact
- These changes standardize the mask tensor shape for better integration with other parts of the codebase, aligning with common practice in machine learning where the batch size (n) comes first.
- By adjusting the bounding box calculation to consider the center of the pixel (`r + 0.5` and `c + 0.5`), the mask cropping becomes more precise, leading to potentially more accurate segmentation results.
- The consistent mask shape and more exact cropping can benefit users by enabling more reliable and potentially more accurate segmentation model training and inference.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@paradigmn interesting, thanks for spotting this! It looks like the SegmentationModel benchmarks reduced mAP slightly after this change though, from 22.39 to 22.33 as shown in 
https://github.com/ultralytics/yolov5/actions/runs/3218894723/jobs/5263602190

It seems like change impacts mAP, so there may be more going on.
Screenshot 2022-10-10 at 15 37 51

**Comment by glenn-jocher** [comment]
@paradigmn I did a before and after on zidane.jpg. Before I see a gap at the top, with this PR I see a gap at the bottom. So it seems like the problem is not really solved.

### Before
zidane

### After
zidane

**Comment by paradigmn** [comment]
@glenn-jocher Yeah, I agree there is still something amiss. I played a bit around and appearently this constallation yields the desired result:
```    
n, h, w = masks.shape
x1, y1, x2, y2 = torch.chunk(boxes[:, :, None], 4, 1)  # x1 shape(1,1,n)
r = torch.arange(w, device=masks.device, dtype=x1.dtype)[None, None, :]  # rows shape(1,1,w)
c = torch.arange(h, device=masks.device, dtype=x1.dtype)[None, :, None]  # cols shape(1,h,1)

return masks * (((r+0.5) >= x1) * (r < x2) * ((c+0.5) >= y1) * (c < y2))
```
I am just not certain, why only the upper left corner requires this transformation. Any idea on this?

**Comment by glenn-jocher** [comment]
@paradigmn got it. I implemented an idea in https://github.com/ultralytics/yolov5/pull/9749/commits/479de10c3b7fc317ae3a2f57ac29999f4e8bffb8, but this dropped mAP even more. I'll try your idea above.

**Comment by glenn-jocher** [comment]
@paradigmn I think one might be especially tricky because the crop_mask() function is used both during training and during inference, so changes we make after training should theoretically always worsen mAP.

**Comment by glenn-jocher** [comment]
@paradigmn new mAP after https://github.com/ultralytics/yolov5/pull/9749/commits/5859180361906158e9741d50f989be78166ac784 is 22.1 vs 22.4 in master and 22.3 for your first change (19.9 with my change).

**Comment by paradigmn** [comment]
@glenn-jocher so how should we continue from here? Should we wait until the weights are retrained again?

**Comment by glenn-jocher** [comment]
@paradigmn see new issue in #10330 related to mask cropping

**Comment by glenn-jocher** [comment]
@paradigmn good news 😃! Your original issue may now be fixed ✅ in PR #10350 by @Laughing-q. This PR improves and speeds up `--retina-mask` mode during inference, i.e. `python segment/predict.py --retina` will plot smoother and more accurate masks at a slower plotting speed than the default masks.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://bit.ly/yolov5-paperspace-notebook">Run on Gradient</a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb">Open In Colab</a> <a href="https://www.kaggle.com/models/ultralytics/yolov5">Open In Kaggle</a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5">Docker Pulls</a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
